extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java

```diff
@@ -0,0 +1,326 @@
+import io.druid.server.namespace.cache.NamespaceExtractionCacheManager;
+import kafka.consumer.ConsumerConfig;
+import kafka.consumer.ConsumerIterator;
+import kafka.consumer.KafkaStream;
+import kafka.consumer.Whitelist;
+import kafka.javaapi.consumer.ConsumerConnector;
+import kafka.message.MessageAndMetadata;
+import kafka.serializer.Decoder;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.UUID;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Pattern;
+
+@JsonTypeName("kafka")
+public class KafkaLookupExtractorFactory implements LookupExtractorFactory
+{
+  private static final Logger LOG = new Logger(KafkaLookupExtractorFactory.class);
+  static final Decoder<String> DEFAULT_STRING_DECODER = new Decoder<String>()
+  {
+    @Override
+    public String fromBytes(byte[] bytes)
+    {
+      return StringUtils.fromUtf8(bytes);
+    }
+  };
+
+  private final Object startStopLock = new Object();
+  private final ListeningExecutorService executorService;
+  private final AtomicLong doubleEventCount = new AtomicLong(0L);
+  private final NamespaceExtractionCacheManager cacheManager;
+  private final String factoryId = UUID.randomUUID().toString();
+  private final AtomicReference<Map<String, String>> mapRef = new AtomicReference<>(null);
+
+  private AtomicBoolean started = new AtomicBoolean(false);
+  private ListenableFuture<?> future = null;
+  private ConsumerConnector consumerConnector;
+
+  @JsonProperty
+  private final String kafkaTopic;
+
+  @JsonProperty
+  private final Map<String, String> kafkaProperties;
+
+  @JsonCreator
+  public KafkaLookupExtractorFactory(
+      @JacksonInject NamespaceExtractionCacheManager cacheManager,
+      @NotNull @JsonProperty(value = "kafkaTopic", required = true) final String kafkaTopic,
+      @NotNull @JsonProperty(value = "kafkaProperties", required = true) final Map<String, String> kafkaProperties
+  )
+  {
+    this.kafkaTopic = Preconditions.checkNotNull(kafkaTopic, "kafkaTopic required");
+    this.kafkaProperties = Preconditions.checkNotNull(kafkaProperties, "kafkaProperties required");
+    executorService = MoreExecutors.listeningDecorator(Execs.singleThreaded(
+        "kafka-factory-" + kafkaTopic + "-%s",
+        Thread.MIN_PRIORITY
+    ));
+    this.cacheManager = cacheManager;
+  }
+
+  public String getKafkaTopic()
+  {
+    return kafkaTopic;
+  }
+
+  public Map<String, String> getKafkaProperties()
+  {
+    return kafkaProperties;
+  }
+
+  @Override
+  public boolean start()
+  {
+    synchronized (startStopLock) {
+      if (started.get()) {
+        LOG.warn("Already started, not starting again");
+        return false;
```

Comment:
In most cases yes, From the API it does not say "if the lookup extractor is started" it says "if start successfully started"